addons/account/models/account_move.py

```diff
@@ -6006,69 +6005,183 @@ def is_internal_partner(partner):
+    def _fix_attachments_on_invoice(self, attachments):
+        """ Fix the given attachments so that only the ones that should be attached to the invoice `self`
+        are actually attached.
+        """
+        self.ensure_one()
+        attachments_to_attach = attachments.filtered(self._should_attach_to_invoice)
+        if attachments_to_attach:
+            attachments_to_attach.write({
+                'res_model': 'account.move',
+                'res_id': self.id,
+            })
+        other_attachments = attachments - attachments_to_attach
+        if attachments_to_unattach := other_attachments.filtered(lambda a: a.res_model == 'account.move' and not a.res_field):
+            attachments_to_unattach.write({
+                'res_model': False,
+                'res_id': 0,
+            })
+
+    def _should_attach_to_invoice(self, attachment):
+        """ Indicate whether a given incoming attachment should be displayed in the invoice's attachments. """
+        return not attachment.res_field and attachment.mimetype in {
+            'text/csv',
+            'application/pdf',
+            'application/vnd.ms-excel',
+            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
+            'application/vnd.oasis.opendocument.spreadsheet',
+            'application/msword',
+            'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
+        }
+
+    def _group_mail_alias_attachments(self, attachments):
+        """ Heuristic to dispatch incoming attachments into groups that should be attached to different records.
+
+            :return: A list of ir.attachment groups (each group is a recordset), each group corresponding to a single invoice.
+        """
+        # 1. Sort the attachments by root attachment type, those without type coming last.
+        attachments = attachments.sorted(lambda a: ((a.root_attachment_id or a).import_type is False, (a.root_attachment_id or a).import_type))
+
+        # 2. Dispatch the attachments into groups.
+        groups = []
+        for attachment in attachments:
+            self._assign_attachment_to_group(attachment, groups)
+
+        # 3. Convert each group from a list of ir.attachment records into an ir.attachment recordset.
+        groups = list(itertools.starmap(self.env['ir.attachment'].union, groups))
+
+        return groups
+
+    def _assign_attachment_to_group(self, incoming_attachment, groups=[]):
+        """ This method is called in a loop to group attachments.
+            This method assumes that the attachments come in an order where they are already sorted by type.
+            Can be overriden to implement a custom grouping behaviour.
+
+            :param incoming_attachment: An ir.attachment record coming from a mail alias
+            :groups: The groups that already exist, each group is a list of ir.attachment records.
+        """
+        def get_root_attachment(attachment):
```

Comment:
My 2 cents: you always have to deal with root_attachment_id like: `a.root_attachment_id or a`
 
What about renaming it as `origin_attachment_id` and instead having a non-stored computed field `root_attachment_id = origin_attachment_id or self` ?

This would make the code much more linear.

Reply:
I really like the idea, thanks!